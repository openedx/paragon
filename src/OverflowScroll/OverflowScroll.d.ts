import React from 'react';
import { OverflowScrollItems } from './OverflowScrollItems';

export interface OverflowScrollProps {
    ariaLabel: string;
    children: React.ReactNode;
    childQuerySelector?: string;
    hasInteractiveChildren?: boolean;
    disableScroll?: boolean;
    disableOpacityMasks?: boolean;
    onScrollPrevious?: () => void;
    onScrollNext?: () => void;
}

declare const OverflowScroll: React.FC<OverflowScrollProps> & {
    Items: React.FC<OverflowScrollItems>;
};

export default OverflowScroll;
