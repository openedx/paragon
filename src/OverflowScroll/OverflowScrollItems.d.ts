import React from 'react';

export interface OverflowScrollItemsProps {
  children: React.ReactNode;
}

declare function OverflowScrollItems(props: OverflowScrollItemsProps): JSX.Element;

export default OverflowScrollItems;
