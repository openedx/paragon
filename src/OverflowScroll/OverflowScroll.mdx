---
title: 'OverflowScroll'
type: 'component'
components:
- OverflowScroll
- OverflowScrollStartSentinel
- OverflowScrollEndSentinel
categories:
- Layout
status: 'New'
designStatus: 'Done'
devStatus: 'Done'
notes: ''
---

A set of React components that encapsulates the logic from [`useOverflowScroll`](./useoverflowscroll) for building carousel-like UI components:
* `OverflowScroll`
* `OverflowScroll.StartSentinel`
* `OverflowScroll.EndSentinel`

## Sample Usage

The following example demonstrates how to use `OverflowScroll`, `OverflowScroll.StartSentinel`, and `OverflowScroll.EndSentinel` to build carousel-like UI components.

```jsx live
() => {
  const ExampleItem = ({ className }) => (
    <div
      className={classNames('example-item border flex-shrink-0 text-center', className)}
      style={{ width: 160 }}
    >
      Item
    </div>
  );

  const OverflowScrollItems = () => {
    const { overflowRef } = useContext(OverflowScrollContext);
    return (
      <div
        ref={overflowRef}
        className="d-flex pb-3"
        style={{ paddingLeft: 1, paddingRight: 1 }}
      >
        <OverflowScroll.StartSentinel />
        {items}
        <OverflowScroll.EndSentinel />
      </div>
    );
  };

  const OverflowScrollControls = () => {
    const {
      isScrolledToStart,
      isScrolledToEnd,
      isOverflowContainerVisible,
      scrollToPrevious,
      scrollToNext,
    } = useContext(OverflowScrollContext);

    return (
      <div className="mb-3">
        <Button
          onClick={scrollToPrevious}
          disabled={isScrolledToStart || !isOverflowContainerVisible}
          size="sm"
          className="mr-2"
        >
          Previous
        </Button>
        <Button
          onClick={scrollToNext}
          disabled={isScrolledToEnd || !isOverflowContainerVisible}
          size="sm"
        >
          Next
        </Button>
      </div>
    );
  };

  const itemCount = 24;
  const items = useMemo(() => Array.from({ length: itemCount }).map((index) => {
    if (index !== itemCount - 1) {
      return <ExampleItem key={uuidv4()} className="mr-2" />;
    }
    // last element, no right margin
    return <ExampleItem key={uuidv4()} />;
  }), []);

  return (
    <OverflowScroll childQuerySelector=".example-item">
      <OverflowScrollControls />
      <OverflowScrollItems />
    </OverflowScroll>
  );
};
```
