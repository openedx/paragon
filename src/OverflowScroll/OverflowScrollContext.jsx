import { createContext } from 'react';

const OverflowScrollContext = createContext();

export default OverflowScrollContext;
