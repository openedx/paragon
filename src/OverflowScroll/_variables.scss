.pgn__overflow-scroll-overflow-container {
  --pgn-overflow-scroll-opacity-mask-transparent: rgba(0, 0, 0, .4);
}

$overflow-scroll-opacity-mask-transparent: var(--pgn-overflow-scroll-opacity-mask-transparent) !default;
