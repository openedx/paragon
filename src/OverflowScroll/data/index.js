export * from './constants';

export { default as useOverflowScroll } from './useOverflowScroll';
export { default as useOverflowScrollActions } from './useOverflowScrollActions';
export { default as useOverflowScrollElementAttributes } from './useOverflowScrollElementAttributes';
export { default as useOverflowScrollEventListeners } from './useOverflowScrollEventListeners';
export { default as useOverflowScrollItems } from './useOverflowScrollItems';
