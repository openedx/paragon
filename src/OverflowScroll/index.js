export { default as OverflowScroll } from './OverflowScroll';
export { default as OverflowScrollContext } from './OverflowScrollContext';
export {
  useOverflowScroll,
  useOverflowScrollItems,
} from './data';
