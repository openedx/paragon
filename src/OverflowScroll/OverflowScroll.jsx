import React, { useState, useMemo } from 'react';
import PropTypes from 'prop-types';
import { useOverflowScroll } from './data';
import OverflowScrollContext from './OverflowScrollContext';
import OverflowScrollItems from './OverflowScrollItems';

function OverflowScroll({
  ariaLabel,
  children,
  childQuerySelector,
  hasInteractiveChildren,
  disableScroll,
  disableOpacityMasks,
  onScrollPrevious,
  onScrollNext,
}) {
  const [overflowRef, setOverflowRef] = useState();

  const {
    isScrolledToStart,
    isScrolledToEnd,
    scrollToPrevious,
    scrollToNext,
  } = useOverflowScroll({
    childQuerySelector,
    hasInteractiveChildren,
    disableScroll,
    disableOpacityMasks,
    onScrollPrevious,
    onScrollNext,
    overflowRef,
  });

  const contextValue = useMemo(() => ({
    overflowRef,
    setOverflowRef,
    isScrolledToStart,
    isScrolledToEnd,
    scrollToPrevious,
    scrollToNext,
  }), [
    overflowRef,
    setOverflowRef,
    isScrolledToStart,
    isScrolledToEnd,
    scrollToPrevious,
    scrollToNext,
  ]);

  return (
    <div
      className="pgn__overflow-scroll"
      role="region"
      aria-label={ariaLabel}
    >
      <OverflowScrollContext.Provider value={contextValue}>
        {children}
      </OverflowScrollContext.Provider>
    </div>
  );
}

OverflowScroll.Items = OverflowScrollItems;

OverflowScroll.propTypes = {
  ariaLabel: PropTypes.string.isRequired,
  children: PropTypes.node.isRequired,
  childQuerySelector: PropTypes.string,
  hasInteractiveChildren: PropTypes.bool,
  disableScroll: PropTypes.bool,
  disableOpacityMasks: PropTypes.bool,
  onScrollPrevious: PropTypes.func,
  onScrollNext: PropTypes.func,
};

OverflowScroll.defaultProps = {
  childQuerySelector: undefined,
  disableScroll: false,
  hasInteractiveChildren: false,
  disableOpacityMasks: false,
  onScrollPrevious: undefined,
  onScrollNext: undefined,
};

export default OverflowScroll;
