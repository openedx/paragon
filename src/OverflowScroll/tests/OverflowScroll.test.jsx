import React from 'react';
import renderer from 'react-test-renderer';
import OverflowScroll from '../OverflowScroll';

describe('<OverflowScroll />', () => {
  it('renders OverflowScroll', () => {
    const tree = renderer.create((
      <OverflowScroll ariaLabel="example overflow scroll">
        <OverflowScroll.Items>
          <div>item</div>
          <div>item</div>
          <div>item</div>
          <div>item</div>
          <div>item</div>
        </OverflowScroll.Items>
      </OverflowScroll>
    )).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
