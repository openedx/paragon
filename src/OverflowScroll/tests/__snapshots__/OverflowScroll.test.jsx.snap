// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<OverflowScroll /> renders OverflowScroll 1`] = `
<div
  aria-label="example overflow scroll"
  className="pgn__overflow-scroll"
  role="region"
>
  <div
    className="pgn__overflow-scroll-item"
  >
    item
  </div>
  <div
    className="pgn__overflow-scroll-item"
  >
    item
  </div>
  <div
    className="pgn__overflow-scroll-item"
  >
    item
  </div>
  <div
    className="pgn__overflow-scroll-item"
  >
    item
  </div>
  <div
    className="pgn__overflow-scroll-item"
  >
    item
  </div>
</div>
`;
